pkg/policy/snapshots.go

```diff
@@ -0,0 +1,192 @@
+	nextIdx     int
+	lastIdx     int
+	storedCnt   int
+	prune       func(*Policies) []error
+}
+
+var (
+	snaps     *snapshots // singleton
+	snapsOnce sync.Once
+)
+
+// newSnapshots creates a new snapshot.
+func newSnapshots() *snapshots {
+	return &snapshots{
+		murw:        sync.RWMutex{},
+		lastVersion: 0,
+		snaps:       [maxSnapshots]*snapshot{},
+		nextIdx:     0,
+		lastIdx:     -1, // no snapshots stored
+		storedCnt:   0,
+		prune:       nil,
+	}
+}
+
+func Snapshots() *snapshots {
+	snapsOnce.Do(func() {
+		snaps = newSnapshots()
+	})
+
+	return snaps
+}
+
+// SetPruneFunc sets the prune function to be called by PruneSnapshotsOlderThan
+// and Store (when overwriting a snapshot).
+func (s *snapshots) SetPruneFunc(prune func(*Policies) []error) {
+	s.murw.Lock()
+	defer s.murw.Unlock()
+
+	s.prune = prune
+}
+
+// Store stores a snapshot of the Policies.
+func (s *snapshots) Store(ps *Policies) {
+	s.murw.Lock()
+	defer s.murw.Unlock()
+
+	s.lastVersion++ // new version
+	if s.lastVersion == 0 {
+		logger.Warnw("Policies version has wrapped around, resetting to 1")
+		s.lastVersion++
+	}
+
+	ps.SetVersion(s.lastVersion)
+
+	snap := &snapshot{
+		time:     time.Now(),
+		version:  s.lastVersion,
+		policies: ps,
+	}
+
+	nextSlot := s.snaps[s.nextIdx]
+	if nextSlot != nil {
+		if s.prune == nil {
+			logger.Warnw("prune function not set, snapshot will not be pruned, only overwritten", "version", nextSlot.version)
+		} else {
+			errs := s.prune(nextSlot.policies)
+			for _, err := range errs {
+				logger.Errorw("failed to prune snapshot", "version", nextSlot.version, "error", err)
+			}
+		}
+	}
+	s.snaps[s.nextIdx] = snap
+	s.lastIdx = s.nextIdx
+	s.nextIdx = (s.nextIdx + 1) % maxSnapshots
+
+	if s.storedCnt < maxSnapshots {
+		s.storedCnt++
+	}
+}
+
+// Get returns a snapshot of the Policies at a given version.
+func (s *snapshots) Get(polsVersion uint16) (*Policies, error) {
+	s.murw.RLock()
+	defer s.murw.RUnlock()
+
+	if s.storedCnt == 0 {
+		return nil, errfmt.Errorf("no snapshots stored")
+	}
+
+	// start from the most recent snapshot
+	startIdx := s.lastIdx
+	for i := 0; i < s.storedCnt; i++ {
+		idx := (startIdx - i + maxSnapshots) % maxSnapshots
+		snap := s.snaps[idx]
+		if snap.version == polsVersion {
+			return snap.policies, nil
+		}
+	}
+
+	return nil, errfmt.Errorf("no snapshot found for version %d", polsVersion)
+}
+
+// GetLast returns the most recent snapshot of the Policies.
+func (s *snapshots) GetLast() (*Policies, error) {
+	s.murw.RLock()
+	defer s.murw.RUnlock()
+
+	if s.storedCnt == 0 {
+		return nil, errfmt.Errorf("no snapshots stored")
+	}
+
+	return s.snaps[s.lastIdx].policies, nil
+}
+
+// TODO: call this function periodically
+// PruneSnapshotsOlderThan prunes snapshots older than a given duration.
+func (s *snapshots) PruneSnapshotsOlderThan(d time.Duration) []error {
+	s.murw.Lock()
+	defer s.murw.Unlock()
+
+	if s.storedCnt <= 1 {
+		return nil
+	}
+
+	if s.prune == nil {
```

Comment:
This is a bit odd to me, because it seems the prune function is mandatory, though one can create a snapshot without it, it is optional to use the `Set` method to pass it or not, if the func is mandatory it should be required by the constructor, one should not be able to create a snapshot without it. Wdyt?

Reply:
`SetPruneFunc()` is temporary and deserves a comment about it. Good point!

As the [periodic control (go routine with Ticker)](https://github.com/aquasecurity/tracee/pull/3305/files/0cbe2ab53a11b05a9fd1ce0acbcc2e6146df7047#diff-5938b4408c6b7a00649735759bab2c23686f6275834942ee729903bc77dfbe5aR141-R143) to Call `PruneSnapshotsOlderThan()` has not yet been implemented, `SetPruneFunc()` and the checks regarding the lack of prune were inserted for test purpose. So you're right, we should make the pruner func mandatory, setting it on the snapshots constructor.